File: services/graphql-dms/src/mutation/delivery_tasks.rs
use async_graphql::{Context, InputObject, Object};
use uuid::Uuid;

#[derive(Debug, Clone, InputObject)]
pub struct CreateDeliveryTaskInput {
    pub package_id: Uuid,
    pub delivery_route_id: Uuid,
    pub route_sequence: i32,
    pub delivery_address: String,
    pub recipient_name: Option<String>,
    pub recipient_phone: Option<String>,
    pub delivery_instructions: Option<String>,
    pub estimated_arrival_time: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug, Clone, Default)]
pub struct Mutation;

#[Object(name = "DmsDeliveryTasksMutations")]
impl Mutation {
    async fn create_delivery_task(
        &self,
        ctx: &Context<'_>,
        payload: CreateDeliveryTaskInput,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "insert into dms.delivery_tasks (package_id, delivery_route_id, route_sequence, delivery_address, recipient_name, recipient_phone, delivery_instructions, estimated_arrival_time) values ($1, $2, $3, $4, $5, $6, $7, $8) returning *"
        )
        .bind(payload.package_id)
        .bind(payload.delivery_route_id)
        .bind(payload.route_sequence)
        .bind(payload.delivery_address)
        .bind(payload.recipient_name)
        .bind(payload.recipient_phone)
        .bind(payload.delivery_instructions)
        .bind(payload.estimated_arrival_time)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_package_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        package_id: Uuid,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set package_id = $1 where id = $2 returning *",
        )
        .bind(package_id)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_delivery_route_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        delivery_route_id: Uuid,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set delivery_route_id = $1 where id = $2 returning *",
        )
        .bind(delivery_route_id)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_route_sequence(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        route_sequence: i32,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set route_sequence = $1 where id = $2 returning *",
        )
        .bind(route_sequence)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_delivery_address(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        delivery_address: String,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set delivery_address = $1 where id = $2 returning *",
        )
        .bind(delivery_address)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_recipient_name(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        recipient_name: Option<String>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set recipient_name = $1 where id = $2 returning *",
        )
        .bind(recipient_name)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_recipient_phone(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        recipient_phone: Option<String>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set recipient_phone = $1 where id = $2 returning *",
        )
        .bind(recipient_phone)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_delivery_instructions(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        delivery_instructions: Option<String>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set delivery_instructions = $1 where id = $2 returning *",
        )
        .bind(delivery_instructions)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_estimated_arrival_time(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        estimated_arrival_time: Option<chrono::DateTime<chrono::Utc>>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set estimated_arrival_time = $1 where id = $2 returning *",
        )
        .bind(estimated_arrival_time)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_actual_arrival_time(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        actual_arrival_time: Option<chrono::DateTime<chrono::Utc>>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set actual_arrival_time = $1 where id = $2 returning *",
        )
        .bind(actual_arrival_time)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_delivery_time(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        delivery_time: Option<chrono::DateTime<chrono::Utc>>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set delivery_time = $1 where id = $2 returning *",
        )
        .bind(delivery_time)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_status(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        status: Option<crate::models::enums::DeliveryTaskStatusEnum>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set status = $1 where id = $2 returning *",
        )
        .bind(status)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_failure_reason(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        failure_reason: Option<crate::models::enums::DeliveryFailureReasonEnum>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set failure_reason = $1 where id = $2 returning *",
        )
        .bind(failure_reason)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_delivery_task_attempt_count(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        attempt_count: Option<i32>,
    ) -> async_graphql::Result<crate::models::delivery_tasks::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::delivery_tasks::Model>(
            "update dms.delivery_tasks set attempt_count = $1 where id = $2 returning *",
        )
        .bind(attempt_count)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn remove_delivery_task(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
    ) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;
        let result = sqlx::query("delete from dms.delivery_tasks where id = $1")
            .bind(id)
            .execute(db)
            .await?;
        if result.rows_affected() != 1 {
            return Err(async_graphql::Error::new("Unable to delete delivery task"));
        }
        Ok("Delivery task removed successfully".into())
    }
}

File: src/graphql/mutations/dms/deliveryTasks.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// DELIVERY TASK MUTATIONS
// ============================================================================

/**
 * Creates a new delivery task.
 * @param payload The input data for creating the delivery task.
 */
export const createDeliveryTask = graphql(`
  mutation CreateDeliveryTask($payload: CreateDeliveryTaskInput!) {
    dms {
      createDeliveryTask(payload: $payload) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the package ID of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param packageId The new package ID.
 */
export const updateDeliveryTaskPackageId = graphql(`
  mutation UpdateDeliveryTaskPackageId($id: UUID!, $packageId: UUID!) {
    dms {
      updateDeliveryTaskPackageId(id: $id, packageId: $packageId) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the delivery route ID of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param deliveryRouteId The new delivery route ID.
 */
export const updateDeliveryTaskDeliveryRouteId = graphql(`
  mutation UpdateDeliveryTaskDeliveryRouteId($id: UUID!, $deliveryRouteId: UUID!) {
    dms {
      updateDeliveryTaskDeliveryRouteId(id: $id, deliveryRouteId: $deliveryRouteId) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the route sequence of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param routeSequence The new route sequence.
 */
export const updateDeliveryTaskRouteSequence = graphql(`
  mutation UpdateDeliveryTaskRouteSequence($id: UUID!, $routeSequence: Int!) {
    dms {
      updateDeliveryTaskRouteSequence(id: $id, routeSequence: $routeSequence) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the delivery address of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param deliveryAddress The new delivery address.
 */
export const updateDeliveryTaskDeliveryAddress = graphql(`
  mutation UpdateDeliveryTaskDeliveryAddress($id: UUID!, $deliveryAddress: String!) {
    dms {
      updateDeliveryTaskDeliveryAddress(id: $id, deliveryAddress: $deliveryAddress) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the recipient name of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param recipientName The new recipient name.
 */
export const updateDeliveryTaskRecipientName = graphql(`
  mutation UpdateDeliveryTaskRecipientName($id: UUID!, $recipientName: String) {
    dms {
      updateDeliveryTaskRecipientName(id: $id, recipientName: $recipientName) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the recipient phone number of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param recipientPhone The new recipient phone number.
 */
export const updateDeliveryTaskRecipientPhone = graphql(`
  mutation UpdateDeliveryTaskRecipientPhone($id: UUID!, $recipientPhone: String) {
    dms {
      updateDeliveryTaskRecipientPhone(id: $id, recipientPhone: $recipientPhone) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the delivery instructions of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param deliveryInstructions The new delivery instructions.
 */
export const updateDeliveryTaskDeliveryInstructions = graphql(`
  mutation UpdateDeliveryTaskDeliveryInstructions($id: UUID!, $deliveryInstructions: String) {
    dms {
      updateDeliveryTaskDeliveryInstructions(id: $id, deliveryInstructions: $deliveryInstructions) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the estimated arrival time of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param estimatedArrivalTime The new estimated arrival time.
 */
export const updateDeliveryTaskEstimatedArrivalTime = graphql(`
  mutation UpdateDeliveryTaskEstimatedArrivalTime($id: UUID!, $estimatedArrivalTime: DateTime) {
    dms {
      updateDeliveryTaskEstimatedArrivalTime(id: $id, estimatedArrivalTime: $estimatedArrivalTime) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the actual arrival time of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param actualArrivalTime The new actual arrival time.
 */
export const updateDeliveryTaskActualArrivalTime = graphql(`
  mutation UpdateDeliveryTaskActualArrivalTime($id: UUID!, $actualArrivalTime: DateTime) {
    dms {
      updateDeliveryTaskActualArrivalTime(id: $id, actualArrivalTime: $actualArrivalTime) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the delivery time of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param deliveryTime The new delivery time.
 */
export const updateDeliveryTaskDeliveryTime = graphql(`
  mutation UpdateDeliveryTaskDeliveryTime($id: UUID!, $deliveryTime: DateTime) {
    dms {
      updateDeliveryTaskDeliveryTime(id: $id, deliveryTime: $deliveryTime) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the status of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param status The new status.
 */
export const updateDeliveryTaskStatus = graphql(`
  mutation UpdateDeliveryTaskStatus($id: UUID!, $status: DeliveryTaskStatusEnum) {
    dms {
      updateDeliveryTaskStatus(id: $id, status: $status) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the failure reason of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param failureReason The new failure reason.
 */
export const updateDeliveryTaskFailureReason = graphql(`
  mutation UpdateDeliveryTaskFailureReason($id: UUID!, $failureReason: DeliveryFailureReasonEnum) {
    dms {
      updateDeliveryTaskFailureReason(id: $id, failureReason: $failureReason) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Updates the attempt count of a delivery task.
 * @param id The UUID of the delivery task to update.
 * @param attemptCount The new attempt count.
 */
export const updateDeliveryTaskAttemptCount = graphql(`
  mutation UpdateDeliveryTaskAttemptCount($id: UUID!, $attemptCount: Int) {
    dms {
      updateDeliveryTaskAttemptCount(id: $id, attemptCount: $attemptCount) {
        id
        routeSequence
        deliveryAddress
        recipientName
        recipientPhone
        deliveryInstructions
        estimatedArrivalTime
        actualArrivalTime
        deliveryTime
        status
        failureReason
        attemptCount
        createdAt
        updatedAt
        package {
          id
          packageNumber
        }
        deliveryRoute {
          id
          routeDate
        }
      }
    }
  }
`);

/**
 * Removes a delivery task by its ID.
 * @param id The UUID of the delivery task to remove.
 */
export const removeDeliveryTask = graphql(`
  mutation RemoveDeliveryTask($id: UUID!) {
    dms {
      removeDeliveryTask(id: $id)
    }
  }
`);
